File: layers/shared/shared/graduation.py
"""Functions for evaluating and executing target graduations."""

import logging
from typing import Any, Optional
from uuid import UUID

import asyncpg

from .queries.graduation import GraduationQueries
from .queries.targets import TargetQueries, auto_enroll_in_default_sequence

logger = logging.getLogger(__name__)

# Security: Allowlist of fields that can be used in rule conditions
ALLOWED_ROOT_FIELDS = frozenset(
    {
        "status",
        "lifecycle_stage",
        "email",
        "first_name",
        "last_name",
        "company",
        "phone",
        "title",
        "industry",
        "custom_fields",
        "metadata",
    }
)
MAX_FIELD_DEPTH = 3


class GraduationError(Exception):
    """Base exception for graduation errors."""

    pass


class TargetNotFoundError(GraduationError):
    """Target does not exist or is not accessible."""

    pass


class UnauthorizedError(GraduationError):
    """Target does not belong to the organization."""

    pass


class InvalidRuleError(GraduationError):
    """Rule configuration is invalid."""

    pass


def _get_field_value(target: dict[str, Any], field: str) -> Any:
    """Extract a field value from target using dot notation.

    Examples:
        _get_field_value(target, "status") -> target["status"]
        _get_field_value(target, "metadata.score") -> target["metadata"]["score"]
    """
    parts = field.split(".")

    # Security validation
    if len(parts) > MAX_FIELD_DEPTH:
        raise InvalidRuleError(f"Field path too deep: {field}")

    if parts[0] not in ALLOWED_ROOT_FIELDS:
        raise InvalidRuleError(f"Field not allowed: {parts[0]}")

    for part in parts:
        if part.startswith("_"):
            raise InvalidRuleError(f"Private field access forbidden: {part}")

    # Traverse the path
    current: Any = target
    for part in parts:
        if current is None or not isinstance(current, dict):
            return None
        current = current.get(part)

    return current


def _evaluate_condition(condition: dict[str, Any], target: dict[str, Any]) -> bool:
    """Evaluate a single condition against a target."""
    field = condition.get("field", "")
    operator = condition.get("operator", "eq")
    expected = condition.get("value")

    actual = _get_field_value(target, field)

    # Handle None consistently - exists check
    if operator == "exists":
        return actual is not None

    # For all other operators, None means condition fails
    if actual is None:
        return False

    if operator == "eq":
        return actual == expected
    elif operator == "ne":
        return actual != expected
    elif operator == "gt":
        return (
            isinstance(actual, (int, float))
            and isinstance(expected, (int, float))
            and actual > expected
        )
    elif operator == "gte":
        return (
            isinstance(actual, (int, float))
            and isinstance(expected, (int, float))
            and actual >= expected
        )
    elif operator == "lt":
        return (
            isinstance(actual, (int, float))
            and isinstance(expected, (int, float))
            and actual < expected
        )
    elif operator == "lte":
        return (
            isinstance(actual, (int, float))
            and isinstance(expected, (int, float))
            and actual <= expected
        )
    elif operator == "contains":
        if isinstance(actual, str):
            return isinstance(expected, str) and expected in actual
        elif isinstance(actual, list):
            return expected in actual
        return False
    else:
        raise InvalidRuleError(f"Unknown operator: {operator}")


def evaluate_conditions(
    conditions: list[dict[str, Any]], target: dict[str, Any]
) -> bool:
    """Evaluate all conditions against a target (implicit AND)."""
    if not conditions:
        return False
    return all(_evaluate_condition(c, target) for c in conditions)


async def find_matching_rule(
    conn: asyncpg.Connection,
    org_id: str,
    target: dict[str, Any],
) -> Optional[dict[str, Any]]:
    """Find the first matching graduation rule for a target."""
    target_type_id = target.get("target_type_id")
    if not target_type_id:
        return None

    rules = await GraduationQueries.get_rules_for_target_type(
        conn, org_id, target_type_id
    )

    # Build evaluation context
    eval_target = {
        **target,
        "custom_fields": target.get("custom_fields") or {},
        "metadata": target.get("metadata") or {},
    }

    for rule in rules:
        conditions = rule.get("conditions", [])
        if conditions and evaluate_conditions(conditions, eval_target):
            return rule

    return None


async def graduate(
    conn: asyncpg.Connection,
    org_id: str,
    target_id: UUID,
    destination_type_id: UUID,
    user_id: Optional[UUID] = None,
    rule_id: Optional[UUID] = None,
) -> dict[str, Any]:
    """Graduate a target to a new type.

    This is the core graduation function. It:
    1. Exits all active sequence enrollments
    2. Updates the target's type
    3. Auto-enrolls in default sequences for the new type
    4. Records the graduation event

    Returns the graduation event record.
    """
    async with conn.transaction():
        # Get and lock the target
        target = await conn.fetchrow(
            "SELECT * FROM targets WHERE id = $1 FOR UPDATE",
            target_id,
        )

        if not target:
            raise TargetNotFoundError(f"Target not found: {target_id}")

        if str(target["organization_id"]) != org_id:
            raise UnauthorizedError(
                f"Target {target_id} does not belong to organization"
            )

        source_type_id = target["target_type_id"]

        if source_type_id == destination_type_id:
            raise GraduationError("Target is already in the destination type")

        # Verify destination type exists and belongs to org
        dest_type = await conn.fetchrow(
            "SELECT id FROM target_types WHERE id = $1 AND organization_id = $2",
            destination_type_id,
            org_id,
        )
        if not dest_type:
            raise GraduationError(f"Invalid destination type: {destination_type_id}")

        # 1. Exit all active sequence enrollments
        exit_result = await conn.execute(
            """
            UPDATE sequence_enrollments
            SET status = 'exited', exit_reason = 'graduated', updated_at = NOW()
            WHERE organization_id = $1 AND target_id = $2
              AND status IN ('active', 'paused')
            """,
            org_id,
            target_id,
        )
        exited_count = int(exit_result.split()[-1]) if exit_result else 0
        logger.info(f"Exited {exited_count} enrollments for target {target_id}")

        # 2. Update target type
        await conn.execute(
            "UPDATE targets SET target_type_id = $1, updated_at = NOW() WHERE id = $2",
            destination_type_id,
            target_id,
        )

        # 3. Auto-enroll in default sequences for new type
        await auto_enroll_in_default_sequence(
            conn, org_id, target_id, destination_type_id
        )

        # 4. Record the graduation event
        event = await GraduationQueries.record_graduation(
            conn=conn,
            org_id=org_id,
            target_id=target_id,
            source_target_type_id=source_type_id,
            destination_target_type_id=destination_type_id,
            rule_id=rule_id,
            manual=user_id is not None and rule_id is None,
            triggered_by_user_id=user_id,
        )

        logger.info(
            f"Graduated target {target_id}: {source_type_id} -> {destination_type_id}",
            extra={"org_id": org_id, "rule_id": str(rule_id) if rule_id else None},
        )

        return event


async def evaluate_and_graduate(
    conn: asyncpg.Connection,
    org_id: str,
    target_id: UUID,
) -> Optional[dict[str, Any]]:
    """Evaluate graduation rules for a target and graduate if matched.

    Returns the graduation event if graduated, None if no rules matched.
    Raises GraduationError on failures (does not silently return None).
    """
    target = await TargetQueries.get_by_id(conn, target_id)

    if not target:
        raise TargetNotFoundError(f"Target not found: {target_id}")

    if str(target.get("organization_id")) != org_id:
        raise UnauthorizedError(f"Target {target_id} does not belong to organization")

    rule = await find_matching_rule(conn, org_id, target)

    if not rule:
        return None

    return await graduate(
        conn=conn,
        org_id=org_id,
        target_id=target_id,
        destination_type_id=rule["destination_target_type_id"],
        rule_id=rule["id"],
    )

File: lib/graduation.ts
import { sql } from "@/lib/db";
import { withTransaction } from "@/lib/db";
import type { PoolClient } from "@neondatabase/serverless";

const ALLOWED_ROOT_FIELDS = new Set([
  "status",
  "lifecycle_stage",
  "email",
  "first_name",
  "last_name",
  "company",
  "phone",
  "title",
  "industry",
  "custom_fields",
  "metadata",
]);

const MAX_FIELD_DEPTH = 3;

export class GraduationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GraduationError";
  }
}

export class TargetNotFoundError extends GraduationError {
  constructor(message: string) {
    super(message);
    this.name = "TargetNotFoundError";
  }
}

export class UnauthorizedError extends GraduationError {
  constructor(message: string) {
    super(message);
    this.name = "UnauthorizedError";
  }
}

export class InvalidRuleError extends GraduationError {
  constructor(message: string) {
    super(message);
    this.name = "InvalidRuleError";
  }
}

function getFieldValue(
  target: Record<string, unknown>,
  field: string,
): unknown {
  const parts = field.split(".");

  if (parts.length > MAX_FIELD_DEPTH) {
    throw new InvalidRuleError(`Field path too deep: ${field}`);
  }

  if (!ALLOWED_ROOT_FIELDS.has(parts[0])) {
    throw new InvalidRuleError(`Field not allowed: ${parts[0]}`);
  }

  for (const part of parts) {
    if (part.startsWith("_")) {
      throw new InvalidRuleError(`Private field access forbidden: ${part}`);
    }
  }

  let current: unknown = target;
  for (const part of parts) {
    if (
      current === null ||
      current === undefined ||
      typeof current !== "object"
    ) {
      return null;
    }
    current = (current as Record<string, unknown>)[part];
  }

  return current ?? null;
}

export function evaluateCondition(
  target: Record<string, unknown>,
  condition: { field: string; operator: string; value?: unknown },
): boolean {
  const { field = "", operator = "eq", value: expected } = condition;
  const actual = getFieldValue(target, field);

  if (operator === "exists") {
    return actual !== null;
  }

  if (actual === null) return false;

  switch (operator) {
    case "eq":
      return actual === expected;
    case "ne":
      return actual !== expected;
    case "gt":
      return (
        typeof actual === "number" &&
        typeof expected === "number" &&
        actual > expected
      );
    case "gte":
      return (
        typeof actual === "number" &&
        typeof expected === "number" &&
        actual >= expected
      );
    case "lt":
      return (
        typeof actual === "number" &&
        typeof expected === "number" &&
        actual < expected
      );
    case "lte":
      return (
        typeof actual === "number" &&
        typeof expected === "number" &&
        actual <= expected
      );
    case "contains":
      if (typeof actual === "string") {
        return typeof expected === "string" && actual.includes(expected);
      }
      if (Array.isArray(actual)) {
        return actual.includes(expected);
      }
      return false;
    default:
      throw new InvalidRuleError(`Unknown operator: ${operator}`);
  }
}

export function evaluateRule(
  target: Record<string, unknown>,
  conditions: Array<{ field: string; operator: string; value?: unknown }>,
): boolean {
  if (!conditions || conditions.length === 0) return false;
  return conditions.every((c) => evaluateCondition(target, c));
}

export async function findMatchingRule(
  orgId: string,
  target: Record<string, unknown>,
): Promise<Record<string, unknown> | null> {
  const targetTypeId = target.target_type_id;
  if (!targetTypeId) return null;

  const rules = await sql`
    SELECT * FROM graduation_rules
    WHERE organization_id = ${orgId}
      AND source_target_type_id = ${targetTypeId}
      AND enabled = TRUE
    ORDER BY created_at
  `;

  const evalTarget: Record<string, unknown> = {
    ...target,
    custom_fields:
      (target.custom_fields as Record<string, unknown>) ?? {},
    metadata: (target.metadata as Record<string, unknown>) ?? {},
  };

  for (const rule of rules) {
    const conditions = rule.conditions as Array<{
      field: string;
      operator: string;
      value?: unknown;
    }>;
    if (conditions && evaluateRule(evalTarget, conditions)) {
      return rule as Record<string, unknown>;
    }
  }

  return null;
}

export async function graduateTarget(opts: {
  orgId: string;
  targetId: string;
  destinationTypeId: string;
  userId?: string;
  ruleId?: string;
}): Promise<Record<string, unknown>> {
  const { orgId, targetId, destinationTypeId, userId, ruleId } = opts;

  return withTransaction(async (client: PoolClient) => {
    // Get and lock the target
    const targetResult = await client.query(
      "SELECT * FROM targets WHERE id = $1 FOR UPDATE",
      [targetId],
    );
    const target = targetResult.rows[0];

    if (!target) {
      throw new TargetNotFoundError(`Target not found: ${targetId}`);
    }

    if (String(target.organization_id) !== orgId) {
      throw new UnauthorizedError(
        `Target ${targetId} does not belong to organization`,
      );
    }

    const sourceTypeId = target.target_type_id;

    if (sourceTypeId === destinationTypeId) {
      throw new GraduationError("Target is already in the destination type");
    }

    // Verify destination type exists and belongs to org
    const destResult = await client.query(
      "SELECT id FROM target_types WHERE id = $1 AND organization_id = $2",
      [destinationTypeId, orgId],
    );
    if (destResult.rows.length === 0) {
      throw new GraduationError(
        `Invalid destination type: ${destinationTypeId}`,
      );
    }

    // 1. Exit all active sequence enrollments
    await client.query(
      `UPDATE sequence_enrollments
       SET status = 'exited', exit_reason = 'graduated', updated_at = NOW()
       WHERE organization_id = $1 AND target_id = $2
         AND status IN ('active', 'paused')`,
      [orgId, targetId],
    );

    // 2. Update target type
    await client.query(
      "UPDATE targets SET target_type_id = $1, segment_id = NULL, updated_at = NOW() WHERE id = $2",
      [destinationTypeId, targetId],
    );

    // 3. Auto-enroll in default sequences for new type
    const defaultSequences = await client.query(
      `SELECT id FROM sequences
       WHERE organization_id = $1
         AND target_type_id = $2
         AND is_default = TRUE
         AND status = 'active'`,
      [orgId, destinationTypeId],
    );

    for (const seq of defaultSequences.rows) {
      // Check not already enrolled
      const existing = await client.query(
        `SELECT id FROM sequence_enrollments
         WHERE organization_id = $1 AND target_id = $2 AND sequence_id = $3
           AND status IN ('active', 'paused')`,
        [orgId, targetId, seq.id],
      );
      if (existing.rows.length === 0) {
        await client.query(
          `INSERT INTO sequence_enrollments (organization_id, target_id, sequence_id, status)
           VALUES ($1, $2, $3, 'active')`,
          [orgId, targetId, seq.id],
        );
      }
    }

    // 4. Record the graduation event
    const eventResult = await client.query(
      `INSERT INTO graduation_events
         (organization_id, target_id, rule_id, source_target_type_id,
          destination_target_type_id, manual, triggered_by_user_id)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       RETURNING *`,
      [
        orgId,
        targetId,
        ruleId ?? null,
        sourceTypeId,
        destinationTypeId,
        userId != null && ruleId == null,
        userId ?? null,
      ],
    );

    return eventResult.rows[0] as Record<string, unknown>;
  });
}

export async function evaluateAndGraduate(
  orgId: string,
  targetId: string,
): Promise<Record<string, unknown> | null> {
  return withTransaction(async (client: PoolClient) => {
    // Lock the target row to prevent concurrent graduation attempts
    const targetResult = await client.query(
      "SELECT * FROM targets WHERE id = $1 FOR UPDATE",
      [targetId],
    );
    const target = targetResult.rows[0];

    if (!target) {
      throw new TargetNotFoundError(`Target not found: ${targetId}`);
    }

    const targetDict = target as Record<string, unknown>;

    if (String(targetDict.organization_id) !== orgId) {
      throw new UnauthorizedError(
        `Target ${targetId} does not belong to organization`,
      );
    }

    const rule = await findMatchingRule(orgId, targetDict);
    if (!rule) return null;

    // Call graduateTarget which uses its own transaction — but we're already
    // in one here, so call the inner logic directly
    const sourceTypeId = targetDict.target_type_id;
    const destinationTypeId = rule.destination_target_type_id as string;

    if (sourceTypeId === destinationTypeId) {
      throw new GraduationError("Target is already in the destination type");
    }

    // Verify destination type
    const destResult = await client.query(
      "SELECT id FROM target_types WHERE id = $1 AND organization_id = $2",
      [destinationTypeId, orgId],
    );
    if (destResult.rows.length === 0) {
      throw new GraduationError(
        `Invalid destination type: ${destinationTypeId}`,
      );
    }

    // 1. Exit active enrollments
    await client.query(
      `UPDATE sequence_enrollments
       SET status = 'exited', exit_reason = 'graduated', updated_at = NOW()
       WHERE organization_id = $1 AND target_id = $2
         AND status IN ('active', 'paused')`,
      [orgId, targetId],
    );

    // 2. Update target type
    await client.query(
      "UPDATE targets SET target_type_id = $1, segment_id = NULL, updated_at = NOW() WHERE id = $2",
      [destinationTypeId, targetId],
    );

    // 3. Auto-enroll in default sequences
    const defaultSequences = await client.query(
      `SELECT id FROM sequences
       WHERE organization_id = $1 AND target_type_id = $2
         AND is_default = TRUE AND status = 'active'`,
      [orgId, destinationTypeId],
    );

    for (const seq of defaultSequences.rows) {
      const existing = await client.query(
        `SELECT id FROM sequence_enrollments
         WHERE organization_id = $1 AND target_id = $2 AND sequence_id = $3
           AND status IN ('active', 'paused')`,
        [orgId, targetId, seq.id],
      );
      if (existing.rows.length === 0) {
        await client.query(
          `INSERT INTO sequence_enrollments (organization_id, target_id, sequence_id, status)
           VALUES ($1, $2, $3, 'active')`,
          [orgId, targetId, seq.id],
        );
      }
    }

    // 4. Record graduation event
    const eventResult = await client.query(
      `INSERT INTO graduation_events
         (organization_id, target_id, rule_id, source_target_type_id,
          destination_target_type_id, manual, triggered_by_user_id)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       RETURNING *`,
      [
        orgId,
        targetId,
        (rule.id as string) ?? null,
        sourceTypeId,
        destinationTypeId,
        false,
        null,
      ],
    );

    return eventResult.rows[0] as Record<string, unknown>;
  });
}
